component/kafka/group/component.go

```diff
@@ -0,0 +1,425 @@
+	}
+
+	if group == "" {
+		errs = append(errs, errors.New("consumer group is required"))
+	}
+
+	if validation.IsStringSliceEmpty(brokers) {
+		errs = append(errs, errors.New("brokers are empty or have an empty value"))
+	}
+
+	if validation.IsStringSliceEmpty(topics) {
+		errs = append(errs, errors.New("topics are empty or have an empty value"))
+	}
+
+	if proc == nil {
+		errs = append(errs, errors.New("work processor is required"))
+	}
+
+	if len(errs) > 0 {
+		return nil, patronErrors.Aggregate(errs...)
+	}
+
+	defaultSaramaCfg, err := kafka.DefaultSaramaConfig(name)
+	if err != nil {
+		return nil, err
+	}
+
+	cmp := &Component{
+		name:         name,
+		group:        group,
+		brokers:      brokers,
+		topics:       topics,
+		proc:         proc,
+		retries:      defaultRetries,
+		retryWait:    defaultRetryWait,
+		batchSize:    defaultBatchSize,
+		batchTimeout: defaultBatchTimeout,
+		failStrategy: defaultFailureStrategy,
+		saramaConfig: defaultSaramaCfg,
+	}
+
+	for _, optionFunc := range oo {
+		err = optionFunc(cmp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return cmp, nil
+}
+
+// Component is a kafka consumer implementation that processes messages in batch
+type Component struct {
+	name         string
+	group        string
+	topics       []string
+	brokers      []string
+	saramaConfig *sarama.Config
+	proc         kafka.BatchProcessorFunc
+	failStrategy kafka.FailStrategy
+	batchSize    uint
+	batchTimeout time.Duration
+	retries      uint
+	retryWait    time.Duration
+	commitSync   bool
+}
+
+// Run starts the consumer processing loop to process messages from Kafka.
+func (c *Component) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	return c.processing(ctx)
+}
+
+func (c *Component) processing(ctx context.Context) error {
+	var componentError error
+
+	retries := int(c.retries)
+	for i := 0; i <= retries; i++ {
+		handler := newConsumerHandler(ctx, c.name, c.group, c.proc, c.failStrategy, c.batchSize,
+			c.batchTimeout, c.commitSync)
+
+		client, err := sarama.NewConsumerGroup(c.brokers, c.group, c.saramaConfig)
+		if err != nil {
+			componentError = err
+			log.Errorf("error creating consumer group client for kafka component: %v", err)
+		}
+
+		if client != nil {
+			// `Consume` should be called inside an infinite loop, when a
+			// server-side rebalance happens, the consumer session will need to be
+			// recreated to get the new claims
+			if err := client.Consume(ctx, c.topics, handler); err != nil {
+				log.Errorf("error from kafka consumer: %v", err)
+			}
+
+			// check if context was cancelled or deadline exceeded, signaling that the consumer should stop
+			if ctx.Err() != nil {
+				log.Infof("kafka component terminating: context cancelled or deadline exceeded")
+				break
+			}
+
+			err = client.Close()
+			if err != nil {
+				log.Errorf("error closing kafka consumer: %v", err)
+			}
+		}
+
+		consumerErrorsInc(c.name)
+		if c.retries > 0 {
```

Comment:
To not wait on last iteration:
```suggestion
		if retries > 0 && i != retries {
```